### PR TITLE
Implemented Searching News by Keyword (Use Case #9)

### PR DESCRIPTION
**Pull request by gummy-bear1**
Implements Use Case 9: Search News by Keyword. Users can enter one or more keywords and retrieve matching articles from the NewsData API. The results appear in the standard article list layout.

Main Flow
- User enters a keyword or phase into search field
- Controller sends the keyword to the Interactor
- Interactor searches the NewsData API
- Presenter updates the ViewModel with the results
- UI displays the matching articles

Alternative Flows
- No results: Shows “No articles found.”
- Empty input: Shows a message asking the user to enter a keyword
- API error: Keeps previous articles and displays an error popup

Additional Notes
- Added full Clean Architecture pipeline for searching
- Minor code cleanup and CheckStyle fixes

### CODE REVIEW

File: src/main/java/data_access/DBUserDataAccessObject.java [hunk 1]
```diff
@@ -60,4 +62,63 @@ public List<Article> fetchTopHeadlines() {
...
+                    + "q=" + keyword
+                    + "&language=en"
+                    + "&removeduplicate=1"
+                    + "&apikey=" + API_KEY;
+
+            Request request = new Request.Builder()
+                    .url(url)
+                    .build();
+
+            Response response = client.newCall(request).execute();
```
Comment (avantika10) [inline src/main/java/data_access/DBUserDataAccessObject.java]: The Response object should be handled in a try block to make sure it's always closed

File: src/main/java/data_access/DBUserDataAccessObject.java [hunk 2]
```diff
@@ -60,4 +62,63 @@ public List<Article> fetchTopHeadlines() {
...
+                        UUID.randomUUID().toString(),
+                        title,
+                        a.optString("description", ""),
+                        a.optString("link", ""),
+                        a.optString("image_url", ""),
+                        source
+                ));
+            }
+        } catch (Exception e) {
+            System.err.println("Error searching news: " + e.getMessage());
```
Comment (avantika10) [inline src/main/java/data_access/DBUserDataAccessObject.java]: throw a custom DataAccessException or return a Result type instead (would make debugging and testing easier)

File: src/main/java/interface_adapter/ViewModel.java [hunk 3]
```diff
@@ -3,37 +3,70 @@
...
     public void setState(T state) {
         this.state = state;
     }
 
+    /**
+     * Fires a property changed event for the state property.
+     */
     public void firePropertyChange() {
         support.firePropertyChange("state", null, this.state);
+        this.support.firePropertyChange("state", null, this.state);
```
Comment (avantika10) [inline src/main/java/interface_adapter/ViewModel.java]: The line 53 and 54 are identical they will trigger all listeners twice for every update. Remove the duplicate line and keep only one call.

**Review by avantika10 — COMMENTED** [review]
The implementation is heading in a great direction and the core functionality works, but there are some improvements needed to ensure reliability and full Clean Architecture compliance.

### **Key areas of feedback:**

**Resource Safety:**
The Response object from OkHttp should be wrapped in a try block (or try-with-resources) to guarantee it is always closed. This prevents resource leaks and makes the data-access code safer.

**Exception Handling:**
Instead of throwing generic exceptions, consider creating a custom DataAccessException or using a Result/Outcome type. This will make debugging easier and allows the use case layer to properly react to failure scenarios.

**Duplicate Listener Calls:**
Lines 53 and 54 currently call the update listeners twice, which will trigger duplicate UI updates. Remove the duplicated line so listeners run exactly once per update.

**Naming & Readability:**
There is a small typo in a method name (excute → execute) which should be corrected for clarity and consistency.

**Null Safety:**
The code checks for a null keyword but not for a null SearchNewsInputData object. Without this check, the controller could trigger a null pointer exception if it ever passes a null input.

**Missing Test Coverage:**
Additional tests are needed for important edge cases:

- Null input object
- DAO returning null
- Case-insensitive keyword matching
- Articles with missing fields (e.g., null title or description)
